etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/ruleengine/transformation/custom/MapLocalObservationCodes.java

```diff
@@ -0,0 +1,152 @@
+
+    public MapLocalObservationCodes() {
+        initMap();
+    }
+
+    @Override
+    public void transform(FhirResource<?> resource, Map<String, String> args) {
+        var bundle = (Bundle) resource.getUnderlyingResource();
+        var observations = HapiHelper.resourcesInBundle(bundle, Observation.class);
+
+        for (Observation obv : observations.toList()) {
+            var codingList = obv.getCode().getCoding();
+
+            if (codingList.size() == 1) {
+                var coding = codingList.get(0);
+
+                var cwe =
+                        coding.getExtensionByUrl(HapiHelper.EXTENSION_CWE_CODING)
+                                .getValue()
+                                .toString();
+
+                // Alt coding is HL7 OBX-3.4,5,6
+                if (Objects.equals(cwe, "alt-coding")
+                        && coding.getSystem().equals(HapiHelper.LOCAL_CODE_URL)) {
+
+                    // Look up the code in the hash map
+                    var identifier = codingMap.get(coding.getCode());
+
+                    if (identifier == null) {
+                        // The local code was not found in the mapping
+                        var msh41Identifier = HapiHelper.getMSH4_1Identifier(bundle);
+                        var msh41Value =
+                                msh41Identifier != null ? msh41Identifier.getValue() : null;
+
+                        LOGGER.logWarning(
+                                "Unmapped local code detected: '{}', from sender: '{}', message Id: '{}'",
+                                coding.getCode(),
+                                msh41Value,
+                                HapiHelper.getMessageControlId(bundle));
+                        continue;
+                    }
+
+                    // Create a new coding source and add it to the coding list
+                    var mappedCoding =
+                            new Coding(
+                                    urlForCodeType(identifier.codingSystem()),
+                                    identifier.code(),
+                                    identifier.display());
+                    mappedCoding.addExtension(
+                            HapiHelper.EXTENSION_CWE_CODING, new StringType("coding"));
+
+                    mappedCoding.addExtension(
+                            HapiHelper.EXTENSION_CODING_SYSTEM,
+                            new StringType(identifier.codingSystem()));
+
+                    codingList.add(0, mappedCoding);
+                }
+            }
+        }
+    }
+
+    private String urlForCodeType(String code) {
+        return switch (code) {
+            case HapiHelper.LOINC_CODE -> HapiHelper.LOINC_URL;
+            case HapiHelper.PLT_CODE -> null;
+            default -> HapiHelper.LOCAL_CODE_URL;
+        };
+    }
+
+    private void initMap() {
+        this.codingMap = new HashMap<String, IdentifierCode>();
+        codingMap.put(
+                "99717-32",
```

Comment:
I'd extract these identifier codes out into an enum so that if they get re-used and updated we don't have multiple places to change the code.

Reply:
I think I'd prefer if we do constants or leave as a HashMap until we move the config to the yaml or db.

Reply:
I think these local codes only belong to UCSD. For that reason, we took the simple approach of having the map as part of the class. If other clients need to add their local codes as well, then that's when we can probably inject the map or enum to the class. But since the codes pertain to only one client, I think we can get away with leaving it as is. Thoughts?

Reply:
Can we put a comment on this map so that other future devs know that these are all UCSD only values?

Reply:
Done